dev-tools/omdb/src/bin/omdb/db/saga.rs

```diff
@@ -0,0 +1,182 @@
+use clap::Subcommand;
+use diesel::prelude::*;
+use nexus_db_model::Saga;
+use nexus_db_model::SagaNodeEvent;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db;
+use nexus_db_queries::db::DataStore;
+use nexus_db_queries::db::datastore::DataStoreConnection;
+use nexus_db_queries::db::datastore::SQL_BATCH_SIZE;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::pagination::paginated;
+use tabled::Tabled;
+use uuid::Uuid;
+
+/// `omdb db saga` subcommand
+#[derive(Debug, Args, Clone)]
+pub struct SagaArgs {
+    #[command(subcommand)]
+    command: SagaCommands,
+}
+
+#[derive(Debug, Subcommand, Clone)]
+enum SagaCommands {
+    /// List running sagas
+    Running,
+
+    /// Inject an error into a saga's currently running node
+    Fault(SagaFaultArgs),
+}
+
+#[derive(Clone, Debug, Args)]
+struct SagaFaultArgs {
+    saga_id: Uuid,
+}
+
+impl SagaArgs {
+    pub async fn exec(
+        &self,
+        omdb: &Omdb,
+        opctx: &OpContext,
+        datastore: &DataStore,
+    ) -> Result<(), anyhow::Error> {
+        match &self.command {
+            SagaCommands::Running => cmd_sagas_running(opctx, datastore).await,
+
+            SagaCommands::Fault(args) => {
+                let token = omdb.check_allow_destructive()?;
+                cmd_sagas_fault(opctx, datastore, args, token).await
+            }
+        }
+    }
+}
+
+async fn cmd_sagas_running(
+    _opctx: &OpContext,
+    datastore: &DataStore,
+) -> Result<(), anyhow::Error> {
+    let conn = datastore.pool_connection_for_tests().await?;
+
+    let sagas =
+        get_all_sagas_in_state(&conn, steno::SagaCachedState::Running).await?;
+
+    #[derive(Tabled)]
+    struct SagaRow {
+        id: Uuid,
+        creator_id: Uuid,
+        time_created: String,
+        name: String,
+        state: String,
+    }
+
+    let rows: Vec<_> = sagas
+        .into_iter()
+        .map(|saga: Saga| SagaRow {
+            id: saga.id.0.into(),
+            creator_id: saga.creator.0,
+            time_created: datetime_rfc3339_concise(&saga.time_created),
+            name: saga.name,
+            state: format!("{:?}", saga.saga_state),
+        })
+        .collect();
+
+    let table = tabled::Table::new(rows)
+        .with(tabled::settings::Style::psql())
+        .to_string();
+
+    println!("{}", table);
+
+    Ok(())
+}
+
+async fn cmd_sagas_fault(
+    _opctx: &OpContext,
+    datastore: &DataStore,
+    args: &SagaFaultArgs,
+    _destruction_token: DestructiveOperationToken,
+) -> Result<(), anyhow::Error> {
+    let conn = datastore.pool_connection_for_tests().await?;
+
+    // Find the most recent node for a given saga
+    let most_recent_node: SagaNodeEvent = {
+        use db::schema::saga_node_event::dsl;
+
+        dsl::saga_node_event
+            .filter(dsl::saga_id.eq(args.saga_id))
+            .order(dsl::event_time.desc())
+            .limit(1)
+            .first_async(&*conn)
+            .await?
+    };
```

Comment:
I think you want to enumerate all nodes for which you have an ActionStarted but no ActionDone or ActionFailed.

I say this because:

- There might be more than one of them.
- They might not be the most recent ones.

Reply:
done in b1108a38a